Fix sliceviewer colorbar autoscale

gemmaguest opened the pull request:
This PR fixes two issues that are causing autoscale to fail for the sliceviewer colorbar:

- Move the `canvas.draw` to before calculating the colorbar limits, because the limits are based on the image data. For Workspace2D this is not recalculated before we redraw.
- For MD workspaces, add a new function to get the image data clipped to the axes extents (because the default `get_array` function always returns the full extent). Note that this works regardless of where the draw is done with the previous fix.

Fixes #33214, fixes #32359

**To test:**

See the instructions in the comment by Richard in the linked issue. Check that autoscaling now works as expected when you zoom in and out. 

Release notes to follow


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

gemmaguest committed:
Fix colorbar autoscale issues

1. Move the canvas.draw to before calculating the colorbar limits,
because the limits are based on the image data. For matrix workspaces
this is not recalculated before we redraw.

2. For MD workspaces, add a new function to get the image data clipped
to the axes extents (the default get_array function always returns
the full extent).

Re #33214

gemmaguest commented:
@RichardWaiteSTFC Sorry for the force push. I've just:
- added release notes 
- simplified the fudge factor for finding pixel edges
- removed a bit of refactoring I did to `extract_matched_slices` which is no longer necessary

RichardWaiteSTFC commented:
> @RichardWaiteSTFC Sorry for the force push. I've just:
> 
> * added release notes
> * simplified the fudge factor for finding pixel edges
> * removed a bit of refactoring I did to `extract_matched_slices` which is no longer necessary

No worries, is it ready for review now? No rush!

RichardWaiteSTFC reviewed (commented):
Just done some testing on Worksapce2D functionality and found a couple of things:
1. It looks like the original colorbar limit is set from the data (all spectra and bins), but if you toggle the autoscale then it sets it using the sampled/interpolated image - it would be good if it was always set to the latter.
![image](https://user-images.githubusercontent.com/55979119/150121229-ee1d06fb-34ec-46e0-8d43-48468a1e35cc.png)
2. This is minor but resizing the sliceviewer window and toggling autoscale sometimes gives a different color limit (can be hard to reproduce) - this makes sense because the sampled image is different, but maybe we should update the colorbar on resizing if autoscale is on? Don't feel strongly about this though.

RichardWaiteSTFC commented:
Functional testing on MDEvent and MDHisto workspaces 2D-4D looks good!

RichardWaiteSTFC commented on Framework/PythonInterface/mantid/plots/modest_image/modest_image.py:
```diff
@@ -87,6 +87,25 @@ def get_array(self):
+        x1 = max(int(np.floor(ind1[0] + 0.5)) + 1, 0)
+
+        # Clip the data to the extents
+        data = self._full_res[y0:y1:1, x0:x1:1]
```
Really nit-picking but we don't need to specify a step of 1 when slicing - happy to leave as is though.

RichardWaiteSTFC commented on qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py:
```diff
@@ -295,7 +295,15 @@ def update_clim_validator(self):
+        data = self.colorbar.mappable.get_array_clipped_to_bounds()
+
+        # If any dimension is zero then we have no data in the display area
+        if any(map(lambda dim: dim == 0, data.shape)):
```
This is a really good addition, I believe it works for MD Histo workspaces (I think it would work for MDEvent anyway becasue BinMD get recalled even for limits outside the extent of the data but it's hard to confirm for MDHisto workspaces as autoscale previously didn't work at all!).
However, I've found that in a Workspace2D (e.g. `MAR11060.raw` which has a spike in the last bin of every spectrum which makes this bug easier to reproduce) if you zoom outside the extent of the data the color limits don't seem to update to (cmax, cmin) = (0,0) it gets frozen at whatever the limits were when there was still data on the image. Or maybe there is a bug elsewhere?

gemmaguest replied:
I've moved the check to the masked data rather than the original data which seems to fix this.

RichardWaiteSTFC reviewed (changes requested):
This is a big improvement and really good tests have been added for the modest image and sliceviewer view. It works very well for MDHisto and MDEvent objects but I have found a couple of issues with Workspace2D - which maybe indicates some edge cases for which there are no tests? I think they are quite minor so if you want to address these in a separate PR then that's fine be me and I will approve.

gemmaguest committed:
Improve handling of empty data, and minor tidying

Check the masked data rather than the original data for empty
dimensions. This catches cases that were previously being missed.

Simplify slicing statement by removing unnecessary step value.

Re #33214

gemmaguest commented:
> Just done some testing on Worksapce2D functionality and found a couple of things:
> 
>     1. It looks like the original colorbar limit is set from the data (all spectra and bins), but if you toggle the autoscale then it sets it using the sampled/interpolated image - it would be good if it was always set to the latter.
>        ![image](https://user-images.githubusercontent.com/55979119/150121229-ee1d06fb-34ec-46e0-8d43-48468a1e35cc.png)
> 
>     2. This is minor but resizing the sliceviewer window and toggling autoscale sometimes gives a different color limit (can be hard to reproduce) - this makes sense because the sampled image is different, but maybe we should update the colorbar on resizing if autoscale is on? Don't feel strongly about this though.

I think both of these would be fixed by updating the colorbar on resize. I've had a look at implementing that but I think it means adding a new signal and I think it's probably more work/risk than it's worth for this release.

RichardWaiteSTFC commented:
> > Just done some testing on Worksapce2D functionality and found a couple of things:
> > ```
> > 1. It looks like the original colorbar limit is set from the data (all spectra and bins), but if you toggle the autoscale then it sets it using the sampled/interpolated image - it would be good if it was always set to the latter.
> >    ![image](https://user-images.githubusercontent.com/55979119/150121229-ee1d06fb-34ec-46e0-8d43-48468a1e35cc.png)
> > 
> > 2. This is minor but resizing the sliceviewer window and toggling autoscale sometimes gives a different color limit (can be hard to reproduce) - this makes sense because the sampled image is different, but maybe we should update the colorbar on resizing if autoscale is on? Don't feel strongly about this though.
> > ```
> 
> I think both of these would be fixed by updating the colorbar on resize. I've had a look at implementing that but I think it means adding a new signal and I think it's probably more work/risk than it's worth for this release.

Ah yes that would make sense! Agreed adding a new signal should wait till the next release - I will make an issue for this.

gemmaguest commented:
> Ah yes that would make sense! Agreed adding a new signal should wait till the next release - I will make an issue for this.

Great, thanks.

This should be ready for re-testing now

RichardWaiteSTFC commented on qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py:
```diff
@@ -296,23 +296,19 @@ def _autoscale_clim(self):
-                self.cmax_value = masked_data.max()
+
+                # If any dimension is zero then we have no data in the display area
+                data_is_empty = any(map(lambda dim: dim == 0, masked_data.shape))
```
This works nicely now - I note that this doesn't cause an issue for log colorscale as when you get limits set to (0,0) then it switches the scale to linear.

RichardWaiteSTFC reviewed (approved):
Looks good, thanks for that last fix and tidy up!
I found one other issue with Workspace2D - which is that we should update the color axis limits when transposing data (for the same reason as we want to update them when resizing the window) - I will add this to issue #33355

martyngigg reviewed (approved):
Nice set of fixes and great to see big increase in test coverage
